backends/cuda-gen/ceed-cuda-gen-operator-build.cpp

```diff
@@ -758,213 +38,251 @@ extern "C" int CeedCudaGenOperatorBuild(CeedOperator op) {
+    ierr = CeedFree(&tensor_basis_kernel_source); CeedChkBackend(ierr);
+  }
+  {
+    char *cuda_gen_template_path, *cuda_gen_template_source;
+    ierr = CeedGetJitAbsolutePath(ceed,
+                                  "ceed/jit-source/cuda/cuda-gen-templates.h",
+                                  &cuda_gen_template_path); CeedChkBackend(ierr);
+    CeedDebug256(ceed, 2, "----- Loading Cuda-Gen Template Source -----\n");
+    ierr = CeedLoadSourceToBuffer(ceed, cuda_gen_template_path, &cuda_gen_template_source);
+    CeedChkBackend(ierr);
+    code << cuda_gen_template_source;
+    ierr = CeedFree(&cuda_gen_template_path); CeedChkBackend(ierr);
+    ierr = CeedFree(&cuda_gen_template_source); CeedChkBackend(ierr);
   }
 
-  code << devFunctions;
+  string q_function_source(qf_data->q_function_source);
+  string q_function_name(qf_data->q_function_name);
+  string operator_name;
+  operator_name = "CeedKernel_Cuda_gen_" + q_function_name;
 
-  string qFunction(qf_data->qFunctionSource);
-  string qFunctionName(qf_data->qFunctionName);
-  string oper;
-  oper = "CeedKernel_Cuda_gen_" + qFunctionName;
-
-  // Find dim and Q1d
-  bool useCollograd = false;
+  // TODO: put in a function?
+  // Find dim and Q_1d
+  bool use_collograd_parallelization = false;
   // Only use collocated gradient algorithm when we actually compute a gradient.
-  if ( dim == 3 ) {
-    for (CeedInt i = 0; i < numinputfields; i++) {
-      ierr = CeedQFunctionFieldGetEvalMode(qfinputfields[i], &emode);
-      if (emode == CEED_EVAL_GRAD) {
-        useCollograd = true;
+  if (dim == 3) {
+    for (CeedInt i = 0; i < num_input_fields; i++) {
+      ierr = CeedQFunctionFieldGetEvalMode(qf_input_fields[i], &eval_mode);
+      if (eval_mode == CEED_EVAL_GRAD) {
+        use_collograd_parallelization = true;
       }
     }
-    for (CeedInt i = 0; i < numoutputfields; i++) {
-      ierr = CeedQFunctionFieldGetEvalMode(qfoutputfields[i], &emode);
-      if (emode == CEED_EVAL_GRAD) {
-        useCollograd = true;
+    for (CeedInt i = 0; i < num_output_fields; i++) {
+      ierr = CeedQFunctionFieldGetEvalMode(qf_output_fields[i], &eval_mode);
+      if (eval_mode == CEED_EVAL_GRAD) {
+        use_collograd_parallelization = true;
       }
     }
   }
-  data->maxP1d = 0;
-  for (CeedInt i = 0; i < numinputfields; i++) {
-    ierr = CeedOperatorFieldGetBasis(opinputfields[i], &basis); CeedChkBackend(ierr);
+  data->max_P_1d = 0;
+  for (CeedInt i = 0; i < num_input_fields; i++) {
+    ierr = CeedOperatorFieldGetBasis(op_input_fields[i], &basis); CeedChkBackend(ierr);
     if (basis != CEED_BASIS_COLLOCATED) {
       ierr = CeedBasisGetData(basis, &basis_data); CeedChkBackend(ierr);
-      ierr = CeedQFunctionFieldGetEvalMode(qfinputfields[i], &emode);
+      ierr = CeedQFunctionFieldGetEvalMode(qf_input_fields[i], &eval_mode);
       CeedChkBackend(ierr);
 
       // Check for collocated gradient
-      useCollograd = useCollograd && basis_data->d_collo_grad_1d; 
```

Comment:
I think this check should be moved to the block checking for collograd and it should only happen with a grad eval mode. I'm seeing ex1/ex2 using the non collograd path when they should and I'm not sure why.

Reply:
I don't have any opinions about where this check goes, but I'm not sure I understand how it could be affecting what you're seeing -- if `d_collo_grad_1d` doesn't exist for the basis, then it can't use collograd, anyway...seems like harmless check in this case?
Do ex1/ex2 never use collograd for dim = 3?  For which parameters do you see the problem?

Reply:
Logging out the value of `use_collograd_parallelization` showed that it was never set to true due to a minor ordering issue. I simplified in the latest push. I was checking with `./build/ex1-volume -ceed /gpu/cuda/gen -d 3`